### PR TITLE
protocol: harden OPEN and UPDATE validation

### LINKED ISSUE #31: [RFC 6793 / major] AS4_PATH / AS4_AGGREGATOR inbound reconstruction + AGGREGATOR attribute
## Context
On the receive side BGPLite never parses or merges the AS4_PATH (type 17) and AS4_AGGREGATOR (type 18) tunnelling attributes defined by RFC 6793, nor does it read/write the base AGGREGATOR (type 7) attribute.

## What's wrong
- `BGPLite.Server/BgpSession.cs:401-419` — `HandleUpdateAsync` switches only on ORIGIN/AS_PATH/NEXT_HOP/COMMUNITY; attributes 17 and 18 are ignored.
- `BGPLite.Protocol/AttributeHelper.cs` — no AS4_PATH / AS4_AGGREGATOR codec; AGGREGATOR (type 7) not read/written.

## What the RFC says
- RFC 6793 §6 — on a 4-byte session, when the received AS_PATH contains AS_TRANS(23456) placeholders plus an AS4_PATH attribute, the speaker MUST reconstruct the full AS_PATH by replacing AS_TRANS entries with the corresponding AS4_PATH segments.
- RFC 6793 §7 — AGGREGATOR (type 7) carries a 4-octet ASN on 4-byte sessions; AS4_AGGREGATOR (type 18) carries the 4-octet aggregator when AGGREGATOR is AS_TRANS.

## Impact
Severity is informational for a leaf route-server that does not re-advertise learned routes (the inbound AS_PATH is not used for transit), but protocol correctness and any future re-advertisement depend on it.

## Fix
- Add AS4_PATH(17) and AS4_AGGREGATOR(18) to the inbound UPDATE switch.
- When the peer advertised capability 65 AND attr 17 is present, reconstruct AS_PATH per RFC 6793 §6 (reuse `ReadAsPath`).
- Read/write AGGREGATOR (type 7) with 4-octet ASN on 4-byte sessions.
- Treat a malformed AS4_PATH as if it were absent (fall back to AS_PATH).
- Round-trip tests with AS_TRANS placeholders.

## Acceptance
- Inbound AS_PATH with AS_TRANS + AS4_PATH reconstructs to the true 4-byte sequence.
- AGGREGATOR round-trips; malformed AS4_PATH degrades gracefully.

Refs: FIXPLAN 7.3; RFC 6793 §6, §7.

### LINKED ISSUE #32: [RFC 4271 / major] Received UPDATE not validated for mandatory path attributes (§5 / §6.3)
## Context
A received UPDATE is accepted even when it is missing one of the well-known mandatory attributes (ORIGIN, AS_PATH, NEXT_HOP), as long as it carries NLRI.

## What's wrong
`BGPLite.Server/BgpSession.cs:399-420` — `HandleUpdateAsync` iterates attributes in a switch and silently uses defaults when ORIGIN/AS_PATH/NEXT_HOP are absent (origin=Incomplete, nextHop=0, asPath=[]). There is no check that all three mandatory attributes were present before the NLRI is accepted into the RouteTable.

## What the RFC says
- RFC 4271 §5 — ORIGIN, AS_PATH, NEXT_HOP are well-known mandatory.
- RFC 4271 §6.3 — an UPDATE that is missing a required well-known attribute is a UPDATE Message Error (subcode 3, "Missing Well-Known Attribute", with the missing type code in Data).

## Impact
A malformed/malicious peer UPDATE with NLRI but no mandatory attributes is accepted; routes with a `0` next-hop and empty AS_PATH enter the RouteTable.

## Fix
- Track a boolean per mandatory attribute while iterating the switch.
- After the attribute loop, if `update.Nlri.Count > 0` and any of ORIGIN/AS_PATH/NEXT_HOP was not seen, send NOTIFICATION(3, subcode "Missing Well-Known Attribute") and drop the UPDATE.

## Acceptance
- A UPDATE with NLRI but missing a mandatory attribute is rejected with NOTIFICATION(3) instead of being accepted.
- Test covers the three missing-attribute cases.

Refs: FIXPLAN 7.4; RFC 4271 §5, §6.3.

### LINKED ISSUE #33: [RFC 5492 / major] No Unsupported Capability NOTIFICATION (OPEN subcode 7)
## Context
BGPLite parses and negotiates capabilities, but never sends an `Unsupported Capability` NOTIFICATION when it receives a capability it cannot tolerate, and does not validate malformed capability TLVs strictly.

## What's wrong
- `BGPLite.Server/BgpSession.cs` (OPEN validation) — no path that raises OPEN error subcode 7 ("Unsupported Capability") with the offending capability TLV in the Data field.
- `BGPLite.Protocol/BgpMessageReader.cs` capability parsing tolerates unknown codes silently rather than signalling per RFC 5492 §4.3 when the capability is non-tolerable.

## What the RFC says
- RFC 5492 §4.3 — a speaker MAY send NOTIFICATION with OPEN Message Error / subcode 7 ("Unsupported Capability") whose Data field contains the unsupported capability; the peer can then retry without it.
- Malformed capability TLV (e.g. Cap Length exceeds remaining bytes) MUST NOT corrupt parsing.

## Fix
- Add an `UnsupportedCapability` NOTIFICATION path (subcode 7) carrying the offending capability TLV in Data.
- Strictly validate capability TLV lengths; a malformed TLV raises the error instead of being silently dropped.

## Acceptance
- Receiving a non-tolerable capability can produce NOTIFICATION(6, 7, <capability>).
- Malformed capability TLV is rejected, not silently ignored.

Refs: FIXPLAN 7.5; RFC 5492 §4.3, §5.

### PR DESCRIPTION
**Pull request by bashrusakh**
## Summary
Hardens BGP OPEN/UPDATE handling in the protocol layer:

- reconstructs RFC 6793 `AS4_PATH` / `AS4_AGGREGATOR` on receive
- validates mandatory UPDATE attributes before accepting NLRI
- sends `OPEN` `Unsupported Capability` notifications for malformed 4-octet ASN capability data

## Closes
- Closes #31 — `AS4_PATH` / `AS4_AGGREGATOR` reconstruction
- Closes #32 — UPDATE missing mandatory attributes
- Closes #33 — Unsupported Capability NOTIFICATION

## Why
Malformed OPEN/UPDATE inputs should fail loudly and consistently in the shared validation path instead of being accepted silently.

## Validation
```bash
dotnet test "BGPLite.Tests/BGPLite.Tests.csproj"
```

## Risk
- Narrow protocol-only change in `BGPLite.Protocol` and `BgpSession`
- Valid peer behavior is unchanged; the new behavior only affects malformed protocol inputs

## Summary by CodeRabbit

* **New Features**
  * Added `NotificationData` support to notification exceptions, enabling optional payload propagation in NOTIFICATION messages.
  * Improved handling of AS4-related path/aggregator attributes and additional well-known attribute validation.
* **Bug Fixes**
  * Tightened AS_PATH/AS4_PATH parsing to reject truncation and ensure full payload consumption.
  * Enhanced UPDATE and OPEN error handling with more precise sub-errors for malformed capabilities and reconstruction failures.
* **Tests**
  * Refined RFC 6793 tests to focus on AS4_PATH round-tripping, reconstruction edge cases, mandatory attribute validation, and malformed capability TLVs.

### CODE REVIEW

File: FIXPLAN.md [hunk 1]
```diff
@@ -0,0 +1,456 @@
...
+2.5 (attribute len) ─┘
+
+3.1 (community filter) ──┐
+3.2 (Route immutable)  ──┤── 5.3 (тесты фильтра)
+3.3 (AddOrUpdate)      ──┤
+3.4 (prefix masking)   ──┤
+3.5 (LPM)              ──┘
+
+4.1-4.5 (config) ── 5.6 (тесты конфига)
+```
```
[the remark this answers is not in the document]
Comment (bashrusakh) [reply FIXPLAN.md:393]: no MD in commits
